# SI 330: Data Manipulation 
## 17 - SQL

### Dr. Chris Teplovs, School of Information, University of Michigan
<small><a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a>This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License</a>.

# <FONT COLOR="red">NOTE: Please install the ```pymysql``` module (conda install pymysql)

# SQL Outline

## Day 1

* AWS Overview
  * EC2, RDS

* Tasks
  * log into AWS Console
  * create EC2 instance
    * yum install mysql
    * wget .sql
    
## Day 2

* RDS
* MySQL

* Tasks
  * create RDS instance
  * load SQL file from EC2
  * browse with MySQLWorkbench
  
## Day 3

* pymysql

* Tasks
  * query goodreads
  
## Day 4

* BigQuery, Homework


## SQL

In general, we are going to:
```
SELECT some_fields FROM some_table
WHERE some_condition_exists
[ GROUP BY some_field ]
[ ORDER BY some_field ]
```

For example:
```
SELECT authors FROM books;
```
**Note:** the convention is to use UPPERCASE for SQL reserved words, and lowercase for the specifics of your instance

If we want all the fields:
```
SELECT * FROM books;
```

Adding a condition:
```
SELECT * FROM books
WHERE original_publication_year >= 2008;
```

And you can also group by a field:
```
SELECT original_publication_year,COUNT(*) FROM books
GROUP BY original_publication_year;
```

## Introduction to pymysql



```pymysql``` is a python interface to SQL servers (including MySQL, sqlite, etc.).  There are other interfaces (such as SQLAlchemy), but pymysql is the most straight-forward to work with. (Note caveats that we'll discuss in class.)

In [2]:
import pymysql

Fill in the ```host```, ```user``` and ```password``` fields in the next code block.  The host is the RDS endpoint; user should probably be your uniqname, and password should be whatever you set you database password to.

Run the block to establish a connection to your RDS-based MySQL server:

In [3]:
connection = pymysql.connect(host='si330-gprime-instance.cst7dbvgzgvt.us-east-1.rds.amazonaws.com', # remove this before submitting
                             user='blank', # remove this before submitting
                             password='blank', # remove this before submitting
                             db='goodreads',
                             charset='utf8mb4',
                             cursorclass=pymysql.cursors.DictCursor)

Here's a "pythonic" way to query your database.  THe following is a bit 

In [4]:
with connection.cursor() as cursor:
    cursor.execute("SELECT * FROM books WHERE original_title = 'The Hunger Games'")
    results = cursor.fetchall()
    for result in results:
        print(result['original_title'],result['authors'])

The Hunger Games Suzanne Collins


### COUNT

In [5]:
with connection.cursor() as cursor:
    cursor.execute("SELECT COUNT(*) FROM books WHERE authors = 'Suzanne Collins'")
    result = cursor.fetchone()
    print(result)

{'COUNT(*)': 9}


We can rename awkward column names by using the ```AS``` operator:

In [6]:
with connection.cursor() as cursor:
    cursor.execute("SELECT COUNT(*) AS cnt FROM books WHERE authors = 'Suzanne Collins'")
    result = cursor.fetchone()
    print(result)

{'cnt': 9}


### DISTINCT

Sometimes, we want to know the number of distinct (unique) values of a field.  Here are two ways of doing this:

In [7]:
with connection.cursor() as cursor:
    cursor.execute('SELECT DISTINCT authors FROM books')
    results = cursor.fetchall()
    print(len(results))

4669


In [8]:
with connection.cursor() as cursor:
    cursor.execute('SELECT COUNT(DISTINCT authors) FROM books')
    result = cursor.fetchone()
    print(result)

{'COUNT(DISTINCT authors)': 4669}


### <font color="magenta">Q1: Which of the above two ways of finding distinct values is more efficient?  Why?</font>

The second way is more efficient because it is just returning the count of each distinct author instead of their names which ultimately is last information and will take less time. 

## SQL from pandas

We are going to continue to use our connection object from above in this section.  Pandas plays nicely with pymysql.

In [9]:
import pandas as pd

Just like what we found with read_csv (a relief compared to CSVDictReader), we have a read_sql method that allows us to create a dataframe based on an SQL query:

In [30]:
firstdf = pd.read_sql('SELECT * FROM books',con=connection)
len(df)

136631

In [29]:
firstdf.head()

book_id  goodreads_book_id  best_book_id  work_id  books_count       isbn  \
0        1            2767052       2767052  2792775          272  439023483   
1        2                  3             3  4640799          491  439554934   
2        3              41865         41865  3212258          226  316015849   
3        4               2657          2657  3275794          487   61120081   
4        5               4671          4671   245494         1356  743273567   

         isbn13                       authors  original_publication_year  \
0  9.780440e+12               Suzanne Collins                       2008   
1  9.780440e+12  J.K. Rowling, Mary GrandPré                       1997   
2  9.780320e+12               Stephenie Meyer                       2005   
3  9.780060e+12                    Harper Lee                       1960   
4  9.780740e+12           F. Scott Fitzgerald                       1925   

                             original_title  ... ratings_count  \
0                          The Hunger Games  ...       4780653   
1  Harry Potter and the Philosopher's Stone  ...       4602479   
2                                  Twilight  ...       3866839   
3                     To Kill a Mockingbird  ...       3198671   
4                          The Great Gatsby  ...       2683664   

  work_ratings_count  work_text_reviews_count  ratings_1  ratings_2  \
0            4942365                   155254      66715     127936   
1            4800065                    75867      75504     101676   
2            3916824                    95009     456191     436802   
3            3340896                    72586      60427     117415   
4            2773745                    51992      86236     197621   

   ratings_3  ratings_4  ratings_5  \
0     560092    1481305    2706317   
1     455024    1156318    3011543   
2     793319     875073    1355439   
3     446835    1001952    1714267   
4     606158     936012     947718   

                                           image_url  \
0  https://images.gr-assets.com/books/1447303603m...   
1  https://images.gr-assets.com/books/1474154022m...   
2  https://images.gr-assets.com/books/1361039443m...   
3  https://images.gr-assets.com/books/1361975680m...   
4  https://images.gr-assets.com/books/1490528560m...   

                                     small_image_url  
0  https://images.gr-assets.com/books/1447303603s...  
1  https://images.gr-assets.com/books/1474154022s...  
2  https://images.gr-assets.com/books/1361039443s...  
3  https://images.gr-assets.com/books/1361975680s...  
4  https://images.gr-assets.com/books/1490528560s...  

[5 rows x 23 columns]

Let's work in pandas, and explore wildcards and regular expressions in SQL:

In [13]:
df = pd.read_sql('SELECT * FROM books WHERE original_title = "The Hunger Games"',con=connection)
df

book_id  goodreads_book_id  best_book_id  work_id  books_count       isbn  \
0        1            2767052       2767052  2792775          272  439023483   

         isbn13          authors  original_publication_year    original_title  \
0  9.780440e+12  Suzanne Collins                       2008  The Hunger Games   

   ... ratings_count work_ratings_count  work_text_reviews_count  ratings_1  \
0  ...       4780653            4942365                   155254      66715   

   ratings_2  ratings_3  ratings_4  ratings_5  \
0     127936     560092    1481305    2706317   

                                           image_url  \
0  https://images.gr-assets.com/books/1447303603m...   

                                     small_image_url  
0  https://images.gr-assets.com/books/1447303603s...  

[1 rows x 23 columns]

In SQL, ```%``` means 1 or more wildcard characters; ```?``` means exactly one wildcard character.  So, if we wanted to look for titles that contain "Hunger Games", we could use:

In [14]:
df = pd.read_sql('SELECT * FROM books WHERE original_title LIKE "%Hunger Games%"',con=connection)
df

book_id  goodreads_book_id  best_book_id   work_id  books_count       isbn  \
0        1            2767052       2767052   2792775          272  439023483   
1      507            7938275       7938275  11349083           78  545265355   

         isbn13          authors  original_publication_year  \
0  9.780440e+12  Suzanne Collins                       2008   
1  9.780550e+12  Suzanne Collins                       2010   

             original_title  ... ratings_count work_ratings_count  \
0          The Hunger Games  ...       4780653            4942365   
1  The Hunger Games Box Set  ...        129665             159760   

   work_text_reviews_count  ratings_1  ratings_2  ratings_3  ratings_4  \
0                   155254      66715     127936     560092    1481305   
1                     6429       1015       2521      13438      42932   

   ratings_5                                          image_url  \
0    2706317  https://images.gr-assets.com/books/1447303603m...   
1      99854  https://images.gr-assets.com/books/1360094673m...   

                                     small_image_url  
0  https://images.gr-assets.com/books/1447303603s...  
1  https://images.gr-assets.com/books/1360094673s...  

[2 rows x 23 columns]

More powerfully, we could use regular expressions:

In [15]:
df = pd.read_sql('SELECT * FROM books WHERE original_title REGEXP ".*Hunger Games.*"',con=connection)
df

book_id  goodreads_book_id  best_book_id   work_id  books_count       isbn  \
0        1            2767052       2767052   2792775          272  439023483   
1      507            7938275       7938275  11349083           78  545265355   

         isbn13          authors  original_publication_year  \
0  9.780440e+12  Suzanne Collins                       2008   
1  9.780550e+12  Suzanne Collins                       2010   

             original_title  ... ratings_count work_ratings_count  \
0          The Hunger Games  ...       4780653            4942365   
1  The Hunger Games Box Set  ...        129665             159760   

   work_text_reviews_count  ratings_1  ratings_2  ratings_3  ratings_4  \
0                   155254      66715     127936     560092    1481305   
1                     6429       1015       2521      13438      42932   

   ratings_5                                          image_url  \
0    2706317  https://images.gr-assets.com/books/1447303603m...   
1      99854  https://images.gr-assets.com/books/1360094673m...   

                                     small_image_url  
0  https://images.gr-assets.com/books/1447303603s...  
1  https://images.gr-assets.com/books/1360094673s...  

[2 rows x 23 columns]

In [16]:
df = pd.read_sql('SELECT * FROM books WHERE original_publication_year IN (2008, 2010)',con=connection)
df.head()

book_id  goodreads_book_id  best_book_id  work_id  books_count        isbn  \
0        1            2767052       2767052  2792775          272   439023483   
1       20            7260188       7260188  8812783          239   439023513   
2       56            1162543       1162543  2960529          183  031606792X   
3       73            1656001       1656001  3328799          161   316068047   
4       88            6442769       6442769  3364505          178  014241493X   

         isbn13          authors  original_publication_year    original_title  \
0  9.780440e+12  Suzanne Collins                       2008  The Hunger Games   
1  9.780440e+12  Suzanne Collins                       2010        Mockingjay   
2  9.780320e+12  Stephenie Meyer                       2008     Breaking Dawn   
3  9.780320e+12  Stephenie Meyer                       2008          The Host   
4  9.780140e+12       John Green                       2008       Paper Towns   

   ... ratings_count work_ratings_count  work_text_reviews_count  ratings_1  \
0  ...       4780653            4942365                   155254      66715   
1  ...       1719760            1870748                    96274      30144   
2  ...       1070245            1107709                    44550     100994   
3  ...        749780             777560                    39778      44215   
4  ...        461311             655271                    42717      14105   

   ratings_2  ratings_3  ratings_4  ratings_5  \
0     127936     560092    1481305    2706317   
1     110498     373060     618271     738775   
2     114893     213402     267706     410714   
3      62501     154906     227180     288758   
4      47183     154479     223895     215609   

                                           image_url  \
0  https://images.gr-assets.com/books/1447303603m...   
1  https://images.gr-assets.com/books/1358275419m...   
2  https://images.gr-assets.com/books/1361039438m...   
3  https://images.gr-assets.com/books/1318009171m...   
4  https://images.gr-assets.com/books/1349013610m...   

                                     small_image_url  
0  https://images.gr-assets.com/books/1447303603s...  
1  https://images.gr-assets.com/books/1358275419s...  
2  https://images.gr-assets.com/books/1361039438s...  
3  https://images.gr-assets.com/books/1318009171s...  
4  https://images.gr-assets.com/books/1349013610s...  

[5 rows x 23 columns]

Sometimes, it's nicer to wrap our SQL across multiple lines, so we can use docstrings ("""...""") to create our SQL.

Here's an example of a join using a WHERE statement:

In [17]:
df = pd.read_sql("""
                SELECT books.original_title, ratings.rating 
                FROM books,ratings 
                WHERE books.authors LIKE '%Rowling%' 
                AND books.book_id = ratings.book_id;
                """, con = connection)   

In [18]:
df.shape

(136631, 2)

In [19]:
df.head()

original_title  rating
0  Harry Potter and the Philosopher's Stone       5
1  Harry Potter and the Philosopher's Stone       4
2  Harry Potter and the Philosopher's Stone       4
3  Harry Potter and the Philosopher's Stone       5
4  Harry Potter and the Philosopher's Stone       3

In [20]:
df.sample(10)

original_title  rating
110227     Harry Potter and the Half-Blood Prince       4
38045   Harry Potter and the Order of the Phoenix       1
22773    Harry Potter and the Prisoner of Azkaban       5
97604        Harry Potter and the Deathly Hallows       5
68219     Harry Potter and the Chamber of Secrets       4
102009     Harry Potter and the Half-Blood Prince       5
67300     Harry Potter and the Chamber of Secrets       4
65260     Harry Potter and the Chamber of Secrets       4
19667    Harry Potter and the Philosopher's Stone       5
20115    Harry Potter and the Philosopher's Stone       5

We can leverage the groupby function in pandas to do helpful calculations:

In [24]:
df.groupby(by='original_title')[['rating']].mean()

rating
original_title                                              
                                                    4.588123
Career of Evil                                      4.134006
Complete Harry Potter Boxed Set                     4.659530
Fantastic Beasts and Where to Find Them             3.775901
Fantastic Beasts and Where to Find Them: The Or...  4.384259
Harry Potter Boxed Set Books 1-4                    4.605357
Harry Potter Collection (Harry Potter, #1-6)        4.699571
Harry Potter and the Chamber of Secrets             4.229418
Harry Potter and the Cursed Child, Parts One an...  3.694737
Harry Potter and the Deathly Hallows                4.525941
Harry Potter and the Goblet of Fire                 4.430780
Harry Potter and the Half-Blood Prince              4.443339
Harry Potter and the Order of the Phoenix           4.358697
Harry Potter and the Philosopher's Stone            4.351350
Harry Potter and the Prisoner of Azkaban            4.418732
Harry, A History: The True Story of a Boy Wizar...  4.090909
Hogwarts: An Incomplete and Unreliable Guide        4.219858
Quidditch Through the Ages                          3.729560
Short Stories from Hogwarts of Heroism, Hardshi...  4.329268
Short Stories from Hogwarts of Power, Politics ...  4.173913
The Casual Vacancy                                  3.174991
The Cuckoo's Calling                                3.794902
The Silkworm                                        3.950251
The Tales of Beedle the Bard                        3.828964
Very Good Lives: The Fringe Benefits of Failure...  4.275510

Here's a schematic representation of the goodreads database:

![](assets/goodreadsEER.png)

In [25]:
df.sample(10)

original_title  rating
41938   Harry Potter and the Order of the Phoenix       3
21577    Harry Potter and the Philosopher's Stone       5
125164               The Tales of Beedle the Bard       3
74945         Harry Potter and the Goblet of Fire       5
116305                       The Cuckoo's Calling       4
59418     Harry Potter and the Chamber of Secrets       5
125413               The Tales of Beedle the Bard       3
86057        Harry Potter and the Deathly Hallows       5
40063   Harry Potter and the Order of the Phoenix       4
106934     Harry Potter and the Half-Blood Prince       5

### <font color="magenta">Q2: How many books in our dataset were originally published in 2016?</font>

In [27]:
q2 = pd.read_sql('SELECT COUNT(*) as cnt FROM books WHERE original_publication_year = 2016',con=connection)
q2

cnt
0  198

There are 198 books in the dataset that were originally published in 2016. 

### <font color="magenta">Q3: How many books were published in each year of original publication in our dataset?</font>


In [32]:
firstdf.groupby(by='original_publication_year').size().head()

original_publication_year
-1750    1
-762     1
-750     2
-720     1
-560     1
dtype: int64

### <font color="magenta">Q4: Do you notice anything interesting or strange about your results from the previous question?  Investigate at least one of these anomalies using either an SQL query or pandas filtering. </font>

There appear to be books in the database that come from years before 0 (B.C.). I investigate this in the cell below using SQL. 

In [39]:
q4 = pd.read_sql('SELECT original_title, original_publication_year FROM books\
                 WHERE original_publication_year < 0', con=connection)
q4.head()

original_title  original_publication_year
0                   Ὀδύσσεια                       -720
1                         Ἰλιάς                       -750
2      孫子兵法 [Sūnzi bīngfǎ]                       -500
3                    Πολιτεία                       -380
4  Οἰδίπους Τύραννος                       -430

### <font color="magenta">Q5: Create a dataframe that contains a list of all book titles from 2017 as well as the tag_names associated with each book.</font>

In [42]:
q5 = pd.read_sql('SELECT books.original_title, tags.tag_name\
                  FROM books, tags, books_tags\
                  WHERE books.goodreads_book_id = books_tags.goodreads_book_id\
                  AND books_tags.tag_id = tags.tag_id\
                  AND books.original_publication_year = 2017;', con=connection)
q5.tail()

original_title               tag_name
1095  Norse Mythology          young-adult\r
1096  The Hate U Give          young-adult\r
1097          Caraval  young-adult-fantasy\r
1098          Caraval  young-adult-fiction\r
1099  The Hate U Give  young-adult-fiction\r

# END OF NOTEBOOK: REMEMBER TO STRIP IDETIFYING INFORMATION FROM TOP OF NOTEBOOK!
Please submit both ipynb and html formats of this notebook to Canvas.